services/worker-manager/src/providers/azure.js

```diff
@@ -0,0 +1,447 @@
+          },
+        }));
+      } catch (err) {
+        // we create multiple resources in order to provision a VM
+        // if we catch an error we want to deprovision those resources
+        await this._removeWorker({...providerData});
+        await workerPool.reportError({
+          kind: 'creation-error',
+          title: 'VM Creation Error',
+          description: err.message,
+          extra: err.details,
+        });
+        return;
+      }
+      this.monitor.log.workerRequested({
+        workerPoolId,
+        providerId: this.providerId,
+        workerGroup: this.providerId,
+        workerId: virtualMachine.vmId,
+      });
+      const now = new Date();
+      await this.Worker.create({
+        workerPoolId,
+        providerId: this.providerId,
+        workerGroup: this.providerId,
+        workerId: virtualMachine.vmId,
+        created: now,
+        lastModified: now,
+        lastChecked: now,
+        expires: taskcluster.fromNow('1 week'),
+        state: this.Worker.states.REQUESTED,
+        capacity: cfg.capacityPerInstance,
+        providerData: {
+          ...providerData,
+          registrationExpiry,
+        },
+      });
+    }));
+  }
+
+  async deprovision({workerPool}) {
+    // nothing to do: we just wait for workers to terminate themselves
+  }
+
+  async registerWorker({worker, workerPool, workerIdentityProof}) {
+    const {document} = workerIdentityProof;
+
+    // use the same message for all errors here, so as not to give an attacker
+    // extra information.
+    const error = () => new ApiError('Signature validation error');
+
+    // workerIdentityProof is a signed message
+
+    // We need to check that:
+    // 1. The embedded document was signed with the private key corresponding to the
+    //    embedded public key
+    // 2. The embedded public key has a proper certificate chain back to a trusted CA
+    // 3. The embedded message contains the vmId that matches the worker making the
+    //    worker making the request
+
+    // signature is base64-encoded DER-format PKCS#7 / CMS message
+
+    // decode base64, load DER, extract PKCS#7 message
+    let decodedMessage = Buffer.from(document, 'base64');
+    let message;
+    try {
+      let asn1 = forge.asn1.fromDer(forge.util.createBuffer(decodedMessage));
+      message = forge.pkcs7.messageFromAsn1(asn1);
+    } catch (err) {
+      this.monitor.warning('Error extracting PKCS#7 message', {error: err.toString()});
+      throw error();
+    }
+
+    let content, crt, pem, sig;
+    // get message content, signing certificate, and signature
+    try {
+      // in testing, message.content is empty, so we access the raw ASN1 structure
+      content = message.rawCapture.content.value[0].value;
+      // convert to pem for convenience
+      // there may be more than one certificate, we only look at the first one
+      crt = message.certificates[0];
```

Comment:
Are the remaining certs in here the intermediate(s)? Or do the intermediates need to be fetched as in the example from the docs? (https://docs.microsoft.com/en-us/azure/virtual-machines/windows/instance-metadata-service#validating-that-the-vm-is-running-in-azure)

Reply:
I think MS only provides the one certificate; the format would allow several, though.

Reply:
The intermediates aren't bundled, if there were more certs here they would in theory be additional signing certs for the message. The intermediate cert URL is available via the authorityInfoAccess location extension in the cert.

+1 to what Dustin said, they sign with one cert from what I've seen